src/crimson/os/seastore/journal/segment_allocator.h

```diff
@@ -408,10 +410,15 @@ class RecordSubmitter {
 
   // blocked for rolling or lack of resource
   std::optional<seastar::shared_promise<> > wait_available_promise;
+  // For the time being, there are only two possible sources for the
+  // record submitter to be unavailable:
+  //    1. RecordSubmitter is "FULL" and the current batch needs flush
+  //    2. Segment of the underlying segment allocator is being rolled
+  size_t unavailable_sources = 0;
   bool has_io_error = false;
   // when needs flush but io depth is full,
   // wait for decrement_io_with_flush()
-  std::optional<seastar::promise<> > wait_unfull_flush_promise;
+  std::optional<seastar::shared_promise<> > wait_unfull_flush_promise;
```

Comment:
Hmm, I think we may not need to generalize wait_available with sources, or relax the wait_unfull_flush_promise to be a shared promise, just to correct a corner case.

How do you think of https://github.com/ceph/ceph/pull/50273/commits/9f8bd71815f5a31c54c922a56225bde072dfdf3b ?

I also include you as the author.